mesonpy/_editable.py

```diff
@@ -0,0 +1,153 @@
+import functools
+import importlib.abc
+import os
+import subprocess
+import sys
+import warnings
+
+from types import ModuleType
+from typing import List, Mapping, Optional, Union
+
+
+if sys.version_info >= (3, 9):
+    from collections.abc import Sequence
+else:
+    from typing import Sequence
+
+
+# This file should be standalone!
+# It is copied during the editable hook installation.
+
+
+_COLORS = {
+    'cyan': '\33[36m',
+    'yellow': '\33[93m',
+    'light_blue': '\33[94m',
+    'bold': '\33[1m',
+    'dim': '\33[2m',
+    'underline': '\33[4m',
+    'reset': '\33[0m',
+}
+_NO_COLORS = {color: '' for color in _COLORS}
+
+
+def _init_colors() -> Mapping[str, str]:
+    """Detect if we should be using colors in the output. We will enable colors
+    if running in a TTY, and no environment variable overrides it. Setting the
+    NO_COLOR (https://no-color.org/) environment variable force-disables colors,
+    and FORCE_COLOR forces color to be used, which is useful for thing like
+    Github actions.
+    """
```

Comment:
Whoever decided to standardize on this interface needs to reconsider its life choices. What was wrong with `COLORS=0` and `COLORS=1` ? I guess the `colors.org` domain was taken...

Reply:
Rich uses this, by the way - If you set both, you get styles (like bold and underline), but not color. Probably not the original intent, but an interesting idea.

I also like setting FORCE_COLOR to 3, just in case levels are supported - 8 being 16 colors, 2 being 256, and 3 true color. Not sure anyone has actually done that, but it would be nice. Most libraries just expect it set or truthy.

The original NO_COLOR was written by someone who really, really hated colors and didn't care one bit about colors being forced on, and didn't specify any standard for turning it on. According to the standard, setting `NO_COLOR` to an empty string or false value also disables color (distinguishing between an empty and a missing environment variable is terrible, IMO, especially because TOML doesn't have a null value)

Reply:
I mean, `NO_COLOR` has been around for a bit, we started using `FORCE_COLOR` as its counterpart in https://github.com/pypa/build, we could have a used `COLORS` to but :shrug:

Reply:
FORCE_COLOR is starting to be a standard too, mostly as a counterpart to NO_COLOR, but a lot of tools use it now, like pytest and such. It's best for everyone to use the same var, so you can set it once in a CI file and get it everywhere, rather than having to look up each tool's custom settings.

Reply:
https://imgs.xkcd.com/comics/standards.png

There's a standard for FORCE_COLOR, another standard for NO_COLOR, and a third standard for "check for both CLICOLOR as well as CLICOLOR_FORCE".

There's also exporting `TERM=dumb`, because dumb terminals don't support color codes, and APPNAME_NO_COLOR for when you want 60 environment variables all doing the same thing within a *single* standard.

CMake has CMAKE_COLOR_MAKEFILE if you want colors in cmake, but not in the generated Makefiles?

GCC has GCC_COLORS=, grep has GREP_COLORS=, for when you want to not only specify that colors shall be used but also a format string specifying a mapping of levels to colors.

Reply:
Yeah, I know. It is simply that `NO_COLOR` is backward in all the ways you want to look at it. And I'm surprised that so many folks blindly followed the trend just because someone spent the money to buy a domain name... Getting the general setting from the `COLORS=` environmental variable, with the possibility of having application specific settings with `APPNAME_COLORS=` would have made so much more sense, and is what most of the GNU tools have been doing for ages.

Reply:
Note: despite my slight sarcasm above, I actually do think that application-specific variables makes some sense. If you use one of these global standards, then on some operating systems, tools like `ls` will force on colored output, which is *guaranteed* to break existing scripts that parse the output.

Reply:
Yes, I like having both a global setting and a custom one, just in case you do really want to only turn on mypy, pytest, tox, etc. Most of those do have that. 95% of the time it's fine to set `FORCE_COLOR` in a CI environment and forget about it.

No argument that COLORS/APP_COLORS would have been better. And for the custom app one, you can do whatever you'd like, since the point of it is it's not a standard. :)